### ISSUE TITLE
A better data grid

### ISSUE THREAD

**Issue by billhao** [billhao/redrock-ios#2]
The current data grid is very primitive. A better data grid needs to provide an interface to support the following features:

- Dynamically add/remove rows and columns
- Dynamically change content of the headers (the first row)
- Select a column
- Get selected column

**Comment by Rosstin** [comment]
With that in mind, do we want to...
•Find some kind of iOS framework and tools to implement those features
•Implement them ourselves in order to be able to control more how they interact with the underlying data

**Comment by billhao** [comment]
Good point, Rosstin. When we started this project in Feb, Joel and I actually have gone through this exercise - looking for existing grid components. My impressions was that there are a number of grid components that are available but some of them are not still in active development and most of them if not all are in objective c. So my conclusion was creating a simple but capable grid in swift myself. What is currently in the app is the first step to that.

As for the expected grid component, it is basically an UI component that can be extended in rows or columns without any limit. It will expose certain functions to interact with outside (or the host). Based on a UICollectionView, the grid class would look something like this:

class DataGridViewController: UICollectionViewController {
  // constructor
  init(numberOfRows, numberOfColumns)

  // set a 2D array as data source
  setDataSource(dataTable)

  // set an array as header
  setHeader(stringArray)

  // add new columns and rows
  addColumns(numberOfNewColumns)
  addRows(numberOfNewRows)
  
  // insert columns in certain position
  insertColumns(insertPosition, numberOfNewColumns)
  
  // remove columns
  removeColumns(columnIndexes)

  // hide columns
  hideColumns(columnIndexes)

  // set column width
  setColumnWidth(columnWidthPairs)

  // set row height. assume all rows have same height
  setRowHeight(height)

  // event when a user taps on a header or cell
  cellPressed(rowIndex, columnIndex)
}

The grid does not involve in any operations like sort. It is simply a UI component for displaying and interacting with the data table. Tapping a header to sort a column can be handled by the host in the cellPressed event rather than by the grid itself.

**Comment by Rosstin** [comment]
Hey cool, I will think about this, this is a good place to start something.

On Fri, May 8, 2015 at 1:35 PM, Hao Wang <notifications@github.com> wrote:

> Good point, Rosstin. When we started this project in Feb, Joel and I
> actually have gone through this exercise - looking for existing grid
> components. My impressions was that there are a number of grid components
> that are available but some of them are not still in active development and
> most of them if not all are in objective c. So my conclusion was creating a
> simple but capable grid in swift myself. What is currently in the app is
> the first step to that.
>
> As for the expected grid component, it is basically an UI component that
> can be extended in rows or columns without any limit. It will expose
> certain functions to interact with outside (or the host). Based on a
> UICollectionView, the grid class would look something like this:
>
> class DataGridViewController: UICollectionViewController {
> // constructor
> init(numberOfRows, numberOfColumns)
>
> // set a 2D array as data source
> setDataSource(dataTable)
>
> // set an array as header
> setHeader(stringArray)
>
> // add new columns and rows
> addColumns(numberOfNewColumns)
> addRows(numberOfNewRows)
>
> // insert columns in certain position
> insertColumns(insertPosition, numberOfNewColumns)
>
> // remove columns
> removeColumns(columnIndexes)
>
> // hide columns
> hideColumns(columnIndexes)
>
> // set column width
> setColumnWidth(columnWidthPairs)
>
> // set row height. assume all rows have same height
> setRowHeight(height)
>
> // event when a user taps on a header or cell
> cellPressed(rowIndex, columnIndex)
> }
>
> The grid does not involve in any operations like sort. It is simply a UI
> component for displaying and interacting with the data table. Tapping a
> header to sort a column can be handled by the host in the cellPressed event
> rather than by the grid itself.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/billhao/redrock-ios/issues/2#issuecomment-100355420>.
>



-- 
Road Redemption <http://store.steampowered.com/app/300380/> • Rex Rocket
<http://store.steampowered.com/app/288020/> • Queen At Arms
<https://www.facebook.com/QueenAtArms> • Hilton HHonors Straight-to-Room App
<http://news.hiltonworldwide.com/index.cfm/newsroom/detail/27192>
